pkg/consensus/consensus.go

```diff
@@ -216,35 +216,49 @@ func (s *service) eventLoop() {
 
 func (s *service) newPayload() payload.ConsensusPayload {
 	return &Payload{
-		message: new(message),
+		message: &message{
+			stateRootEnabled: s.stateRootEnabled(),
+		},
 	}
 }
 
+// stateRootEnabled checks if state root feature is enabled on current height.
+// It should be called only from dbft callbacks and is not protected by any mutex.
+func (s *service) stateRootEnabled() bool {
+	return s.Chain.GetConfig().EnableStateRoot
+}
+
 func (s *service) newPrepareRequest() payload.PrepareRequest {
-	sr, err := s.Chain.GetStateRoot(s.Chain.BlockHeight())
-	if err != nil {
+	if !s.stateRootEnabled() {
 		return new(prepareRequest)
 	}
-	return &prepareRequest{
-		proposalStateRoot: sr.MPTRootBase,
+	sr, err := s.Chain.GetStateRoot(s.Chain.BlockHeight())
+	if err == nil {
+		return &prepareRequest{
+			stateRootEnabled:  s.stateRootEnabled(),
+			proposalStateRoot: sr.MPTRootBase,
+		}
 	}
+	return &prepareRequest{stateRootEnabled: s.stateRootEnabled()}
```

Comment:
Didn't we already check for `s.stateRootEnabled()` 10 lines above?

Reply:
Fixed.